### ISSUE TITLE
ReadFormAsync continues reading after exception

### ISSUE THREAD

**Issue by yshahin** [dotnet/aspnetcore#17723]
### Describe the bug
ReadFormAsync keeps reading form data after exception thrown for invalid data.

### To Reproduce
Send binary data file to api with content-type "application/x-www-form-urlencoded".
File must be large to see the problem, 10MB file will do it.
Since this is invalid data an exception System.IO.InvalidDataException is thrown.
```
System.IO.InvalidDataException
  HResult=0x80131501
  Message=Form key length limit 2048 exceeded.
  Source=Microsoft.AspNetCore.WebUtilities
  StackTrace:
   at Microsoft.AspNetCore.WebUtilities.FormPipeReader.ThrowKeyTooLargeException() in /_/src/Http/WebUtilities/src/FormPipeReader.cs:line 308
```
The using the code snippet below, the exception will be caught and and 400 sent back.

``` C#
app.Use(async (context, next) =>
{
    try
    {
        // TODO: try-catch log unexpected errors
        await next.Invoke().ConfigureAwait(true);
    }
    catch (InvalidDataException dataEx)
    {
        context.Response.StatusCode = (int)HttpStatusCode.BadRequest;
        await context.Response.WriteAsync("BadRequest").ConfigureAwait(false);
    }
});
```
Since the data is large and ReadFormAsync doesn't stop reading, we get this exception System.InvalidOperationException: 'Reading is already in progress.'.
```
System.InvalidOperationException
  HResult=0x80131509
  Message=Reading is already in progress.
  Source=Microsoft.AspNetCore.Server.Kestrel.Core
  StackTrace:
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1ContentLengthMessageBody.TryReadInternal(ReadResult& readResult) in /_/src/Servers/Kestrel/Core/src/Internal/Http/Http1ContentLengthMessageBody.cs:line 126
```

### Further technical details
- ASP.NET Core version 3
- Include the output of `dotnet --info`
.NET Core SDK (reflecting any global.json):
Version:   3.1.100
Commit:    cd82f021f4
Runtime Environment:
OS Name:     Windows
OS Version:  10.0.18363
OS Platform: Windows
RID:         win10-x64
Base Path:   C:\Program Files\dotnet\sdk\3.1.100\
Host (useful for support):
Version: 3.1.0
Commit:  65f04fb6db
.NET Core SDKs installed:
2.1.202 [C:\Program Files\dotnet\sdk]
2.1.507 [C:\Program Files\dotnet\sdk]
2.1.509 [C:\Program Files\dotnet\sdk]
2.1.701 [C:\Program Files\dotnet\sdk]
2.1.801 [C:\Program Files\dotnet\sdk]
2.1.802 [C:\Program Files\dotnet\sdk]
2.2.401 [C:\Program Files\dotnet\sdk]
3.0.100 [C:\Program Files\dotnet\sdk]
3.1.100 [C:\Program Files\dotnet\sdk]
.NET Core runtimes installed:
Microsoft.AspNetCore.All 2.1.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
Microsoft.AspNetCore.All 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
Microsoft.AspNetCore.All 2.1.13 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
Microsoft.AspNetCore.All 2.2.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
Microsoft.AspNetCore.App 2.1.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
Microsoft.AspNetCore.App 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
Microsoft.AspNetCore.App 2.1.13 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
Microsoft.AspNetCore.App 2.2.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
Microsoft.AspNetCore.App 3.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
Microsoft.AspNetCore.App 3.1.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
Microsoft.NETCore.App 2.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.NETCore.App 2.1.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.NETCore.App 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.NETCore.App 2.1.13 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.NETCore.App 2.2.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.NETCore.App 3.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.NETCore.App 3.1.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.WindowsDesktop.App 3.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
Microsoft.WindowsDesktop.App 3.1.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
- The IDE (VS / VS Code/ VS4Mac) you're running on, and it's version
Microsoft Visual Studio Enterprise 2019
Version 16.3.8
VisualStudio.16.Release/16.3.8+29503.13
Microsoft .NET Framework
Version 4.8.03752

**Comment by Tratcher** [comment]
Duplicate of https://github.com/aspnet/AspNetCore/issues/14727 which is already fixed in 3.1.

**Comment by Tratcher** [comment]
I see you're using the 3.1 SDK, but can you clarify that you're using the 3.0 runtime?

**Comment by yshahin** [comment]
@Tratcher I have 3.1 installed but I am using 3.0 for my project. I will test with 3.1 and reply back

**Comment by yshahin** [comment]
@Tratcher I tested on 3.1 and the problem is still there, it is not resolved.

**Comment by Tratcher** [comment]
@halter73 @jkotalik

**Comment by Tratcher** [comment]
@yshahin can you attach a Fiddler trace of the malformed request?

**Comment by halter73** [comment]
In this case, it looks like the bug is that FormPipeReader.ReadFormAsync doesn't call PipeReader.AdvanceTo() when ParseFormValues throws. The call to AdvanceTo should probably be moved to a finally.

https://github.com/aspnet/AspNetCore/blob/0792e9a4f56de47580c4490c27ca453934552246/src/Http/WebUtilities/src/FormPipeReader.cs#L85-L114

Kestrel doesn't try to put the pipe back into a readable state if middleware calls PipeReader.ReadAsync() without calling PipeReader.AdvanceTo() before exiting. The bug in 3.0 was in HttpRequestStream (i.e. HttpContext.Request.Body) itself.

**Comment by halter73** [comment]
I talked a bit with @Tratcher offline about this. While this is definitely a bug in FormPipeReader, I agree that Kestrel should report this kind of issues more clearly.  Logging an InvalidOperationException with the message "Reading is already in progress." is probably not the best way of telling developers that there's a missing call to Httpcontext.Request.BodyPipe.AdvanceTo().

**Comment by yshahin** [comment]
@halter73 so you are saying there are 2 issues here.
1) The error message is not clear
2) ReadFormAsync should call AdvanceTo when a failure happens, to put the stream in a readable state.

Does this summarize the issues?

**Comment by halter73** [comment]
@yshahin Yes. We also discussed whether Kestrel should put the pipe back in a readable state on the middleware's behalf when the middleware doesn't do so itself. I think giving a clearer error message should be sufficient though.

**Comment by yshahin** [comment]
@halter73, I don't understand. Which middleware are you talking about? The middleware in asp.net core right?
I don't think its kestrels responsibility, the middleware should handle the pipe correctly on failure.

**Comment by anurse** [comment]
@jkotalik can you pop in and add an Advance to fix it? This is likely a patch candidate, but let's get it in master first and then open a PR to backport (and discuss servicing approval there).

**Comment by halter73** [comment]
> I don't think its kestrels responsibility, the middleware should handle the pipe correctly on failure.

I agree, but at the very least Kestrel could provide a clearer error message so middleware authors can more easily understand the problem is the missing call to AdvanceTo.

**Comment by yshahin** [comment]
I am facing a similar issue now, probably the same root cause.
Microsoft.AspNetCore.Server.Kestrel.Core.BadHttpRequestException: Reading the request body timed out due to data arriving too slowly. See MinRequestBodyDataRate.

The client is returning "The stream does not support concurrent IO read or write operations." cause this exception does not put the stream in a readable state.

Can we add this to the checks for the bug fix?

Thanks

**Comment by Tratcher** [comment]
@yshahin that's different from the Form issue. That looks like a Kestrel issue. What version is this from?

**Comment by yshahin** [comment]
@Tratcher this is .net core 3.0

**Comment by Tratcher** [comment]
Ok, please open a separate issue.

**Comment by halter73** [comment]
@yshahin I would also check whether the new issue still exists in 3.1. It's possible that https://github.com/aspnet/AspNetCore/issues/14727 partially fixes your issue if you're seeing an InvalidOperationException claiming "Reading is already in progress" after a request timeout.

As for the request body timeout itself, I would try to verify the client actually did send a complete request body in a reasonable amount of time. Either way, @Tratcher is right that further discussion on this should happen in a new issue since it seems fairly different from the initial issue you filed.

**Comment by yshahin** [comment]
I think its related, my code used CopyTo and when this exception happens
Microsoft.AspNetCore.Server.Kestrel.Core.BadHttpRequestException: Reading the request body timed out due to data arriving too slowly. See MinRequestBodyDataRate.
   at Microsoft.AspNetCore.Server.Kestrel.Core.BadHttpRequestException.Throw(RequestRejectionReason reason)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1ContentLengthMessageBody.ReadAsyncInternal(CancellationToken cancellationToken)
   at System.IO.Pipelines.PipeReader.CopyToAsyncCore[TStream](TStream destination, Func`4 writeAsync, CancellationToken cancellationToken)
   at SpeakerRecognition.Common.Utilities.ControllerExtensions.CopyAudioToAsync(HttpRequest request, MemoryStream audioStream) in D:\a\1\s\private\SpeakerRecognition.Hosts.Service\Extensions\ControllerExtensions.cs:line 177
   at SpeakerRecognition.Hosts.Controllers.IdentificationController.IdentifyAsync(String identificationProfileIds, Boolean shortAudio) in D:\a\1\s\private\SpeakerRecognition.Hosts.Service\Controllers\IdentificationController.cs:line 108
   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.TaskOfIActionResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Awaited|12_0(ControllerActionInvoker invoker, ValueTask`1 actionResultValueTask)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)

The stream is not cleaned correctly and so writing 400 results in IO concurrent issue again, similar to this one.
I have not tried asp.net core 3.1, @halter73  do you know any tool that can simulate slow connections for rest apis?

**Comment by Tratcher** [comment]
The difference is that the bug is in CopyTo rather than in the form reader. It's a similar bug in a different place so we'd like to track the issue separately.

**Comment by yshahin** [comment]
@Tratcher, @halter73 any work around in the mean time?
I tried calling Close and cancelPendingRead but they don't help.
I don't know where to get the AdvanceTo parameter from.

**Comment by yshahin** [comment]
I upgraded to asp.net 3.1 and updated all the nuget packages and I am unable to trigger the issue.
For #18138 I was unable to recreate it locally, it was only triggered in production. But haven't seen the issue after the upgrade.
Not sure if you want to keep this bug open to patch .net 3.0 or you will consider the resolution to upgrade to 3.1?
Thanks

**Comment by Tratcher** [comment]
@yshahin thanks for the update. We'll double check #18138 and close it if we can't reproduce it in 3.1 either.

**Comment by yshahin** [comment]
@Tratcher, just hit the issue today. This is after the 3.1 upgrade.

**Comment by Tratcher** [comment]
@yshahin in ReadFormAsync or CopyTo?

**Comment by yshahin** [comment]
@Tratcher ReadFormAsync

**Comment by yshahin** [comment]
@halter73, I would like to point out that this issue is sever. This problem causes the connection to our proxy to break, which the proxy interprets as 500. Causing SLA misses while the response should be 400. And in a more recent case 404 triggered this case.

**Comment by Tratcher** [comment]
> And in a more recent case 404 triggered this case.

Can you explain that? The identified issue only relates to exception handling, not 404 scenarios.

**Comment by yshahin** [comment]
I am unable to repro this, but for some reason the same issue we see on the proxy happened for a 404 request. It might be another issue. I am still trying to figure out how to repro.

**Comment by halter73** [comment]
> @Tratcher, @halter73 any work around in the mean time?

You can try sending a "Connection: close" response header when you manually send a 400 response. This should prevent the client from sending another request, and will prevent Kestrel from trying to read it which should avoid the issue of connection breaking unexpectedly.

**Comment by anurse** [comment]
Fixed in https://github.com/dotnet/aspnetcore/pull/18939 which will be released in 3.1.3 (which is not the immediate next patch release unfortunately but it's on the way!)
